test/core/utils/qsa.js

```diff
@@ -48,184 +48,229 @@ function getTestDom() {
+          dom = getTestDom();
+
+          // prove we're using the cache by deleting all the children
+          dom[0].children = [];
+        });
+
+        it('should not use the cache if not using the top-level node', function() {
+          var nodes = axe.utils.querySelectorAllFilter(dom, 'ul');
+
+          // this would return 4 nodes if we were still using the
+          // top-level cache
+          var result = axe.utils.querySelectorAllFilter(nodes[0], 'li');
+          assert.equal(result.length, 2);
+        });
+      }
+
+      it('should find nodes using just the tag', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'li');
+        assert.equal(result.length, 4);
+      });
+      it('should find nodes using parent selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'ul > li');
+        assert.equal(result.length, 4);
+      });
+      it('should NOT find nodes using parent selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'div > li');
+        assert.equal(result.length, 0);
+      });
+      it('should find nodes using hierarchical selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, 'div li');
+        assert.equal(result.length, 4);
+      });
+      it('should find nodes using class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.breaking');
+        assert.equal(result.length, 2);
+      });
+      it('should find nodes using hierarchical class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.first .breaking');
+        assert.equal(result.length, 2);
+      });
+      it('should NOT find nodes using hierarchical class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.second .breaking');
+        assert.equal(result.length, 0);
+      });
+      it('should find nodes using multiple class selector', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '.second.third');
+        assert.equal(result.length, 1);
+      });
+      it('should find nodes using id', function() {
+        var result = axe.utils.querySelectorAllFilter(dom, '#one');
+        assert.equal(result.length, 1);
+      });
+
+      // can only select shadow dom nodes if we aren't using the
+      // top-level node
+      if (describeName === 'without cache') {
+        it('should find nodes using id, but not in shadow DOM', function() {
+          var result = axe.utils.querySelectorAllFilter(
+            dom[0].children[0],
+            '#one'
+          );
+          assert.equal(result.length, 1);
+        });
+        it('should find nodes using id, within a shadow DOM', function() {
+          var result = axe.utils.querySelectorAllFilter(
+            dom[0].children[0].children[2],
+            '#one'
+          );
+          assert.equal(result.length, 1);
+        });
+      }
```

Comment:
This is also added

Reply:
Why does this not get tested with the cahce?